rpcs3/Emu/RSX/rsx_methods.cpp

```diff
@@ -1093,19 +1093,58 @@ namespace rsx
 				}
 				else
 				{
-					if (out_pitch != in_pitch || out_pitch != out_bpp * out_w)
+					const bool is_overlapping = scale_x > 0 && scale_y > 0 && [&]() -> bool
 					{
-						for (u32 y = 0; y < out_h; ++y)
+						const u32 src_max = src_offset + (in_pitch * (in_h - 1) + (in_bpp * in_w)) - 1;
+						const u32 dst_max = dst_offset + (out_pitch * (out_h - 1) + (out_bpp * out_w)) - 1;
+						return (src_offset >= dst_offset && src_offset <= dst_max) ||
+						 (dst_offset >= src_offset && dst_offset <= src_max);
+					}();
+
+					if (is_overlapping)
+					{
+						if (out_pitch != in_pitch || out_pitch != out_bpp * out_w)
 						{
-							u8 *dst = pixels_dst + out_pitch * y;
-							u8 *src = pixels_src + in_pitch * y;
+							temp2.resize(out_bpp * out_w * out_h);
+							u8* bufptr = temp2.data(); 
+
+							for (u32 y = 0; y < out_h; ++y, bufptr += out_w * out_bpp)
+							{
+								u8 *src = pixels_src + in_pitch * y;
+
+								std::memcpy(bufptr, src, out_w * out_bpp);
+							}
 
-							std::memmove(dst, src, out_w * out_bpp);
+							bufptr = temp2.data(); 
+
+							for (u32 y = 0; y < out_h; ++y, bufptr += out_w * out_bpp)
+							{
+								u8 *dst = pixels_src + in_pitch * y;
+
+								std::memcpy(dst, bufptr, out_w * out_bpp);
+							}
+						}
+						else
+						{
+							std::memmove(pixels_dst, pixels_src, out_pitch * out_h);
 						}
 					}
 					else
 					{
-						std::memmove(pixels_dst, pixels_src, out_pitch * out_h);
+						if (out_pitch != in_pitch || out_pitch != out_bpp * out_w)
+						{
+							for (u32 y = 0; y < out_h; ++y)
+							{
+								u8 *dst = pixels_dst + out_pitch * y;
```

Comment:
I know licm pass would fix this, but whats with the row pointer computation on every loop? Why not just use an addition like you did with bufptr before? This is more of a consistency thing and also helps when I inevitably have to debug something without optimizations enabled.